app/controllers/v0/messages_controller.rb

```diff
@@ -67,8 +66,55 @@ def thread
 
     private
 
+    def get_folder(folder_id)
+      folders = client.get_folder(folder_id)
+
+      raise VA::API::Common::Exceptions::RecordNotFound, folder_id unless folders.present?
+      folders.data.first
+    end
+
+    # If page is not supplied, we are bringing back all results in the set. In that case
+    # adjust the per_page count to the TBD maximum number of messages allowed per MHV call
+    def pagination_params
+      folder_id = params[:folder_id].try(:to_i)
+      folder = get_folder(folder_id)
```

Comment:
I think you're already refactoring this, but I think pagination_params should not have to be dependent on a GET request. If it's abstracted in the model, the model can recursively make GET requests until it has the full collection, and then you can just use the pagination that collection supports out of the box... so I think all you need is the API method that does the get request, to make multiple GET requests until it has everything and build a collection based off of the merging of each of the individual requests.